arch/x86/x86.c

```diff
@@ -14,6 +14,22 @@ ZLIB_INTERNAL int x86_cpu_has_sse2;
 ZLIB_INTERNAL int x86_cpu_has_sse42;
 ZLIB_INTERNAL int x86_cpu_has_pclmulqdq;
 
+#ifdef _MSC_VER
+#include <intrin.h>
+#define CPU_PROCINFO_AND_FEATUREBITS 1
+
+void ZLIB_INTERNAL x86_check_features(void) {
+	unsigned int registers[4];
+	__cpuid(registers, CPU_PROCINFO_AND_FEATUREBITS);
+
+	unsigned int ecx = registers[2];
+	unsigned int edx = registers[3];
+
+	x86_cpu_has_sse2 = edx & 0x4000000;
```

Comment:
The problem here is that either the code is cleaned up or there is long span of duplicated code that makes it really hard to follow what exactly is compiler-dependent code and what is not... I'm all for documenting things but in the end this is nothing more than work-in-progress as it is missing the part where all the optimizations actually get enabled at build time.